WIP: interpret functions several times and remove cold blocks with deopts

undingen opened the pull request:
This is very hacked together prototype of a JIT which runs the interpreter several times to collect traces.
After JIT_THRESHOLD is reach it will JIT the function but will remove all cold blocks which did not execute yet.
They get replaced to calls to "mydeopt" which will continue to execute the cold path in the interpreter and
continue with tracing.
After JIT_RECOMPILE_THRESHOLD deopts we will JIT the function again.

Currently this is only tested on prime_summing.
And it actually slows it down :-(.
But be disabling the call to createDeopts this patch actually improves the performance because more paths are traced.

Recursive functions are not handles, also it leaks memory and abuses the existing classes to store data...

undingen commented on src/interp2.cpp:
```diff
@@ -1338,19 +1443,30 @@ long _runJitTarget(JitTarget* target, ...) {
 
-    struct timespec start;
-    struct timespec end;
+    target->inside = 0;
```
should be
```suggestion
    target->inside = 1;
```

undingen commented:
I'm currently trying to change it to use `llvm.experimental.guard` / `llvm.experimental.deoptimization` (so stackmaps again) will see if this helps.

undingen commented:
Ok so I got it running with `llvm.experimental.guard` https://github.com/undingen/dpro/commit/4070cbc577f129847a9bf48ae9a0b5f4b352ea2a.
But I had to upgrade to the latest upstream llvm version and had to apply this patch: https://reviews.llvm.org/D59020 (which is quite fresh: last updates yesterday ;))

Also after lowering the guards to deoptimize calls LLVM will not generate return instructions after the call to `__llvm_deoptimize` in the machine code (on purpose as an optimization). So I tried to handle the exit from deopt by c++ exceptions and libunwind unwinding but both approaches failed. I guess llvm also does not generate EH info for the call to `__llvm_deoptimize`?!?
I did not want to patch llvm so instead just before codegen I replace the `llvm.experimental.deoptimize` calls to `mydeopt` (but keep using the `deopt` operandbundle). This prevents llvm from removing the return instruction.

The good news is that this decreases the slowdown this "deopt" approach introduced a lot (30% if I remember correct - don't want to check - compiling llvm takes a while on my laptop).
I have not checked why but I assume it is because llvm has an easier job optimizing a function which contains a lot of  `llvm.experimental.guard` instructions instead of branches?

kmod commented:
I like the "run the interpreter multiple times to generate more traces" change -- I think we could split that out and merge that now if you want!  As I mentioned via email, my worry is that the deopt stuff won't help that much until we start optimizing the hotpath more aggressively and then can benefit from showing that we stayed on it.

Another worry I have about deopt is that prime_summing is particularly type-stable, and I worry that deopt would perform worse on other benchmarks :(  We don't have a ton working at this moment though

undingen commented:
I updated this patch, it's still a hack but it's at least a bit cleaner..

I split out the deopt implementation and the patch which generates deopts for every non executed block into separate commits.
It looks like just generating deopts for all not executed blocks is a bad idea. I hope we can use in the future the deopt functionality but with our own strategy of where to insert guards.

Also before my method to find live values was extremely slow and did not work on nbody for example.
I replaced it now with a modified algorithm from llvm

kmod commented:
Yeah, I think I agree now that deopt-untaken-branches is probably bad.  I was thinking maybe we could have a blacklist of branch types that don't make sense, but I think the whitelist (insert extra guards) approach is definitely better -- I think it's safer, but it's also more powerful, because we can add guards that don't exist as branches in the code.  For example instead of guarding on `x->ob_type->tp_as_number->nb_add`, we guard on `x->ob_type`.

My type_speculation branch (add type speculation guards) is coming along well, I'm excited to merge these two branches together :)

undingen commented:
Ok so i updated the patch again...
I fixed two important problems:
- I did not handle phis correctly
- if we add i1s (any non byte sized type) we trigger an assert inside llvm so i extend them now to i8s

It also tracks now the DeoptInfos we create so that eventually (not yet implemented) we can release the allocated memory and module if we know we don't need it anymore.

I also removed the commit which marks all non executed blocks as deopts.
So right now this PR does nothing but it implements `createDeopts` which allows one to create deopts in the future.

I also updated llvm again because the patch which updates the stackmap parser got merged :-)

undingen commented:
The patch should now be ready for review.
Sorry I made the last few changes in a hurry at the airport gate so I hope I did not screw up something.

kmod commented on src/deopt.h:
```diff
@@ -0,0 +1,44 @@
+
+void createDeopts(llvm::Function *func, _JitTarget& jit_target, llvm::SetVector<llvm::BasicBlock*> deopt_blocks);
+// llvm assumes that __llvm_deoptimize will never return. But we want our deopt function to just return to the caller when it's finished.
+// Therfore replace calls to experimental.deoptimize with a call to our "deopt" function + deopt operand bundles.
```
Sorry I don't quite follow this, the langref [says](https://llvm.org/docs/LangRef.html#llvm-experimental-deoptimize-intrinsic) that it "returns the value returned by the deoptimization continuation".  I'm not very familiar with the new guard/deopt intrinsics though

kmod commented on src/deopt.cpp:
I'm a little bit unsure how to use this interface for adding guards.  Should we have the guard-failure block initially just branch back to the guard-success block, and then turn the failure block into a deopt block?
